File: executor/native/executor/executor.c
#include "executor.h"
#include <liburing.h>
#include <liburing/io_uring.h>
#include <scheduler/scheduler.h>
#include <system/library.h>
#include "configuration.h"
#include "constants.h"
#include "errors.h"
#include "module.h"

struct executor_instance* executor_create(struct executor_configuration* configuration, struct executor_scheduler* scheduler, uint32_t id)
{
    struct executor_instance* executor = executor_module_new(sizeof(struct executor_instance));
    if (executor == NULL)
    {
        event_propagate_local(executor_error_out_of_memory());
        return NULL;
    }

    executor->id = id;
    executor->configuration = *configuration;
    executor->scheduler = scheduler;
    executor->state = EXECUTOR_STATE_PAUSED;

    executor->completions = executor_module_allocate(configuration->ring_size, sizeof(struct io_uring_cqe*));
    if (executor->completions == NULL)
    {
        event_propagate_local(executor_error_out_of_memory());
        return NULL;
    }

    executor->ring = executor_module_new(sizeof(struct io_uring));
    if (executor->ring == NULL)
    {
        event_propagate_local(executor_error_out_of_memory());
        return NULL;
    }

    int32_t result = io_uring_queue_init(configuration->ring_size, executor->ring, configuration->ring_flags);
    if (result)
    {
        event_propagate_local(executor_error_system(-result));
        return NULL;
    }

    executor->descriptor = executor->ring->ring_fd;

    return executor;
}

int16_t executor_register_on_scheduler(struct executor_instance* executor, int64_t callback)
{
    executor->callback = callback;
    executor->state = EXECUTOR_STATE_IDLE;
    struct io_uring_sqe* sqe = io_uring_get_sqe(executor->ring);
    if (unlikely(sqe == NULL))
    {
        event_propagate_local(executor_error_ring_full());
        return MODULE_ERROR_CODE;
    }
    io_uring_prep_msg_ring(sqe, executor->scheduler->descriptor, EXECUTOR_SCHEDULER_REGISTER, (uintptr_t)executor, 0);
    sqe->flags |= IOSQE_CQE_SKIP_SUCCESS;
    io_uring_submit(executor->ring);
    return 0;
}

int16_t executor_unregister_from_scheduler(struct executor_instance* executor, bool stop)
{
    executor->state = stop ? EXECUTOR_STATE_STOPPED : EXECUTOR_STATE_PAUSED;
    struct io_uring_sqe* sqe = io_uring_get_sqe(executor->ring);
    if (unlikely(sqe == NULL))
    {
        event_propagate_local(executor_error_ring_full());
        return MODULE_ERROR_CODE;
    }
    io_uring_prep_msg_ring(sqe, executor->scheduler->descriptor, EXECUTOR_SCHEDULER_UNREGISTER, executor->id, 0);
    sqe->flags |= IOSQE_CQE_SKIP_SUCCESS;
    io_uring_submit(executor->ring);
    return 0;
}

int32_t executor_peek(struct executor_instance* executor)
{
    struct executor_configuration* configuration = &executor->configuration;
    io_uring_submit_and_get_events(executor->ring);
    return io_uring_peek_batch_cqe(executor->ring, &executor->completions[0], configuration->ring_size);
}

int16_t executor_call_native(struct executor_instance* executor, int32_t target_ring_fd, struct executor_task* task)
{
    struct io_uring* ring = executor->ring;
    struct io_uring_sqe* sqe = io_uring_get_sqe(ring);
    if (unlikely(sqe == NULL))
    {
        event_propagate_local(executor_error_ring_full());
        return MODULE_ERROR_CODE;
    }
    task->source = executor->descriptor;
    task->target = target_ring_fd;
    io_uring_prep_msg_ring(sqe, target_ring_fd, EXECUTOR_CALL, (uintptr_t)task, 0);
    sqe->flags |= IOSQE_CQE_SKIP_SUCCESS;
    executor_submit(executor);
    return 0;
}

int16_t executor_callback_to_native(struct executor_instance* executor, struct executor_task* task)
{
    struct io_uring* ring = executor->ring;
    struct io_uring_sqe* sqe = io_uring_get_sqe(ring);
    if (unlikely(sqe == NULL))
    {
        event_propagate_local(executor_error_ring_full());
        return MODULE_ERROR_CODE;
    }
    uint64_t target = task->source;
    task->source = executor->descriptor;
    task->target = target;
    io_uring_prep_msg_ring(sqe, target, EXECUTOR_CALLBACK, (uintptr_t)task, 0);
    sqe->flags |= IOSQE_CQE_SKIP_SUCCESS;
    executor_submit(executor);
    return 0;
}

void executor_destroy(struct executor_instance* executor)
{
    io_uring_queue_exit(executor->ring);
    executor_module_delete(executor->ring);
    executor_module_delete(executor->completions);
    executor_module_delete(executor);
}

int16_t executor_awake_begin(struct executor_instance* executor)
{
    executor->state = EXECUTOR_STATE_WAKING;
    struct io_uring_sqe* sqe = io_uring_get_sqe(executor->ring);
    if (unlikely(sqe == NULL))
    {
        event_propagate_local(executor_error_ring_full());
        return MODULE_ERROR_CODE;
    }
    io_uring_prep_msg_ring(sqe, executor->scheduler->descriptor, EXECUTOR_SCHEDULER_POLL, (uintptr_t)executor, 0);
    sqe->flags |= IOSQE_CQE_SKIP_SUCCESS;
    return 0;
}

void executor_awake_complete(struct executor_instance* executor, uint32_t completions)
{
    io_uring_cq_advance(executor->ring, completions);
    io_uring_submit(executor->ring);
    executor->state = EXECUTOR_STATE_IDLE;
}

File: executor/dart/lib/bindings/executor/executor.dart
// Generated
// ignore_for_file: unused_import

import 'dart:ffi';
import 'package:ffi/ffi.dart';
import '../../executor/bindings.dart';

final class io_uring extends Opaque {}

final class executor_instance extends Struct {
  @Int64()
  external int callback;
  external Pointer<executor_scheduler> scheduler;
  external Pointer<io_uring> ring;
  external Pointer<Pointer<executor_completion_event>> completions;
  external executor_configuration configuration;
  @Int32()
  external int descriptor;
  @Uint32()
  external int id;
  @Int8()
  external int state;
}

@Native<Pointer<executor_instance> Function(Pointer<executor_configuration> configuration, Pointer<executor_scheduler> scheduler, Uint32 id)>(isLeaf: true)
external Pointer<executor_instance> executor_create(Pointer<executor_configuration> configuration, Pointer<executor_scheduler> scheduler, int id);

@Native<Int8 Function(Pointer<executor_instance> executor, Int64 callback)>(isLeaf: true)
external int executor_register_scheduler(Pointer<executor_instance> executor, int callback);

@Native<Int8 Function(Pointer<executor_instance> executor)>(isLeaf: true)
external int executor_unregister_scheduler(Pointer<executor_instance> executor);

@Native<Int32 Function(Pointer<executor_instance> executor)>(isLeaf: true)
external int executor_peek(Pointer<executor_instance> executor);

@Native<Void Function(Pointer<executor_instance> executor)>(isLeaf: true)
external void executor_submit(Pointer<executor_instance> executor);

@Native<Int8 Function(Pointer<executor_instance> executor)>(isLeaf: true)
external int executor_awake_begin(Pointer<executor_instance> executor);

@Native<Void Function(Pointer<executor_instance> executor, Uint32 completions)>(isLeaf: true)
external void executor_awake_complete(Pointer<executor_instance> executor, int completions);

@Native<Int8 Function(Pointer<executor_instance> executor, Int32 target_ring_fd, Pointer<executor_task> message)>(isLeaf: true)
external int executor_call_native(Pointer<executor_instance> executor, int target_ring_fd, Pointer<executor_task> message);

@Native<Int8 Function(Pointer<executor_instance> executor, Pointer<executor_task> message)>(isLeaf: true)
external int executor_callback_to_native(Pointer<executor_instance> executor, Pointer<executor_task> message);

@Native<Int8 Function(Pointer<io_uring> ring, Int32 source_ring_fd, Int32 target_ring_fd, Pointer<executor_task> message)>(isLeaf: true)
external int executor_call_dart(Pointer<io_uring> ring, int source_ring_fd, int target_ring_fd, Pointer<executor_task> message);

@Native<Int8 Function(Pointer<io_uring> ring, Int32 source_ring_fd, Pointer<executor_task> message)>(isLeaf: true)
external int executor_callback_to_dart(Pointer<io_uring> ring, int source_ring_fd, Pointer<executor_task> message);

@Native<Void Function(Pointer<executor_instance> executor)>(isLeaf: true)
external void executor_destroy(Pointer<executor_instance> executor);
